fix(dashboard): Tab menu visible for urls trailing '/'

CuriousCorrelation opened the pull request:
## Description

This bug #10666 was due to this

https://github.com/gitpod-io/gitpod/blob/402894b63b604234899db345fca0a9ce3750cec1/components/dashboard/src/Menu.tsx#L109

failing if `url` contained trailing `'/'`.

Here's a test

``` js
var url = new URL('https://gitpod.io/workspaces/')
>> ["/workspaces"].includes(url.pathname)
<- false
```
So if that was true, it should also happen to 

https://github.com/gitpod-io/gitpod/blob/402894b63b604234899db345fca0a9ce3750cec1/components/dashboard/src/Menu.tsx#L108

which was indeed the case, see https://gitpod.io/new/ (notice trailing '/').

A simple solution would have been to include another element (`"workspaces/")` in the array or discard trailing '/' and check with `includes`. Adding a new element can potentially cause similar problem if we were to forget duplicating for some new path. For the second option, I first thought of regex but it got really unreadable, so instead I chose `split` and `filter`

``` js
url.pathname.split('/').filter(p => p).join('/')
```
This would work but what if we are on a resource but not exactly on base? For example

(Browser console)
``` js
const url1 = new URL('https://gitpod.io/workspaces/testing')
var testingSubpath = url1.pathname.split('/').filter(p => p).join('/')
testingSubpath
"workspaces/testing"
```

``` js
var testing = ["workspaces", "teams/new", "open"]
```
Tab menu shouldn't be displayed in path `workspace/testing` because it is `isWorkspacesUI` (a subdirectory)

``` js
testing.includes(testingSubpath)
false
``` 
A better way would be to test "subpath" against "rootpath"

(Browser console)
``` js
var url1 = new URL("https://gitpod.io/teams/new/testing")
var trimmedPath = url1.pathname.split('/').filter(Boolean).join('/')
trimmedPath
"teams/new/testing"
```
Now if we were to test against 

``` js
var testing = ["workspaces", "teams/new", "open"]
```
we would get an array

``` js
testing.map(r => trimmedPath.startsWith(r))
Array(3) [ false, true, false ]
```
on which we can do an exhaustive search like

``` js
testing.map(r => trimmedPath.startsWith(r)).some(Boolean)
true
```

This would also work for `isAdminUI` check

https://github.com/gitpod-io/gitpod/blob/402894b63b604234899db345fca0a9ce3750cec1/components/dashboard/src/Menu.tsx#L110

(Browser console)
``` js
const url = new URL('https://gitpod.io/admin/testing')
const isAdminUI = url.pathname.startsWith("/admin")
isAdminUI
true
var testing = ["admin"]
var trimmedPath = url.pathname.split('/').filter(Boolean).join('/')
trimmedPath
"admin/testing"
testing.map(r => trimmedPath.startsWith(r)).some(Boolean)
true
```
---
If this subpath check behaviour isn't exactly what we need, we can swap positions and get exact results

``` js
var testing = ["admin"]
testing.map(r => r.startsWith(testingSubpath)).some(Boolean)
false
testing.map(r => testingSubpath.startsWith(r)).some(Boolean)
true
```
## Related Issue(s)
Fixes  #10666

## How to test
Described above

## Release Notes
```release-note
Fixed Tab menu being visible for urls with trailing '/'
```

## Documentation
Does this PR require updates to the documentation at www.gitpod.io/docs?
* No

- [x] /werft with-preview

gtsiolis commented:
/werft run

gtsiolis commented:
/werft run

laushinka commented:
Added werft with-preview in the description

laushinka commented:
Hi @CuriousCorrelation, sorry for the delay in reviewing. The build failed, and to make sure that we have everything up to date, would you mind rebasing to the latest in main? 🙏🏽

CuriousCorrelation commented:
@laushinka No worries at all! Branch should be updated now

laushinka commented:
Thank you for this contribution, @CuriousCorrelation!
It indeed fixed #10666. It doesn't yet resolve #10499. You're very welcome to pick that up for your next PR, or include it in this one :)

Before doing that, would you mind writing a simple test for `inResource()`? It should be very similar and as straightforward as [this test](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/src/App.test.ts).
This will also make it much easier to test the #10499 cases should you or someone else pick that one :)

Thanks again 🙏🏽

CuriousCorrelation commented:
Writing tests would be no trouble at all. I'll try to get it done as soon as possible.

CuriousCorrelation commented:
@laushinka Great idea for adding tests! Not only did it help me fix a couple of edge issues, I believe it also fixed #10499

The problem was the trim function's application on first argument and not on second; the list of resources, because both of them can have a trailing or leading '/'.

A couple of changes from the last commit:

1. Moved `inResource` to `utils.ts`. I figured that'd be a better location since it's not technically a 'component'.
2. Which meant we could add a `utils.test.ts` for placing `inResource`'s tests. Also easier to add more cases.

Let me know if these changed are fine, or if there's something I need to change.

laushinka commented:
> @laushinka Great idea for adding tests! Not only did it help me fix a couple of edge issues, I believe it also fixed #10499

That's awesome, @CuriousCorrelation! I'm on vacation right now, so let me tag my team @gitpod-io/engineering-webapp so someone could take over :) 
Otherwise I will pick this up again next Monday.

laushinka commented:
/werft run

laushinka commented:
Hi @CuriousCorrelation, I'm back from vacation now - sorry that it's taking a long time. 

The build is failing because we're missing a license header for `utils.test.ts`. Would you mind doing the following:
1. Run `leeway run components:update-license-header` in the terminal
2. Rebase to main.

laushinka commented:
/werft run

laushinka commented:
Connection got closed - running this again

/werft run

laushinka commented:
Not sure why builds are failing. Trying again..

/werft run

laushinka commented:
@CuriousCorrelation This looks good! I'm happy to approve this (finally). 

#10499 is not yet fixed, which I believe is due to [this line](https://github.com/gitpod-io/gitpod/blob/04b9a640240d01c243e397ef55bffae6a69c42c7/components/dashboard/src/components/PageWithSubMenu.tsx#L31). All that needs to be done I believe is replacing that check with your new logic 😉

I don't want to block this from being merged any further though because you've done a lot of work. You're very welcome to work on #10499 with the hint above.

Thanks for your contribution 🔥

laushinka commented:
(Oops I should've asked you to squash. Note for next time!)

CuriousCorrelation commented:
> #10499 is not yet fixed, which I believe is due to this line.

Ah makes sense, thanks for the heads up!
